### ISSUE TITLE
Cannot capture the audio of two games with UE4Capture at the same time.

### ISSUE THREAD

**Issue by hwwxj** [adamrehn/UE4Capture#3]
platform: Linux Ubuntu
UE4 version: 4.25
hello, I am trying to capture the audio of two games with UE4Capture at the same time

![image](https://user-images.githubusercontent.com/33884571/109133020-7c0a2380-778f-11eb-8ffa-f6259d75d2b8.png)

but when I start two games, only one IPC is working (the one which starts later)

and when I check the status of soundcard,  only one device is using.

![image](https://user-images.githubusercontent.com/33884571/109131818-1bc6b200-778e-11eb-83a9-518a7c3386bd.png)

where I close the game which starts later, then I can capture the audio of the first game.

May I ask that is UE4Capture support capturing two games at the same time?
@adamrehn

**Comment by adamrehn** [comment]
In theory two independent pairs of producers and consumers should be able to function in parallel if they're using different IPC prefixes. Are your two games completely independent, each with their own separate consumer process? Also, does your system have PulseAudio installed or are you just using ALSA directly?

**Comment by hwwxj** [comment]
> In theory two independent pairs of producers and consumers should be able to function in parallel if they're using different IPC prefixes. Are your two games completely independent, each with their own separate consumer process? Also, does your system have PulseAudio installed or are you just using ALSA directly?

Are your two games completely independent, each with their own separate consumer process?   ---  yes, two independent games with their own separate consumer process.
does your system have PulseAudio installed or are you just using ALSA directly?   using alsa directly.

**Comment by adamrehn** [comment]
Interesting. I've seen instances before where ALSA will lock an audio device for exclusive use and prevent other applications from initialising audio output until the original application is closed, although in my experience that should prevent the second game from ever generating audio at all since the Unreal Engine only attempts to initialise its audio subsystem once during startup. Would you be willing to install PulseAudio and report back whether that makes any difference?

**Comment by hwwxj** [comment]
> Interesting. I've seen instances before where ALSA will lock an audio device for exclusive use and prevent other applications from initialising audio output until the original application is closed, although in my experience that should prevent the second game from ever generating audio at all since the Unreal Engine only attempts to initialise its audio subsystem once during startup. Would you be willing to install PulseAudio and report back whether that makes any difference?

I am using the virtual sound card since there is no hardware soundcard in my server, not sure if this will cause the issue like this.    anyway, I will try the PulseAudio asap, and check whether this will make any difference.

**Comment by hwwxj** [comment]
> Interesting. I've seen instances before where ALSA will lock an audio device for exclusive use and prevent other applications from initialising audio output until the original application is closed, although in my experience that should prevent the second game from ever generating audio at all since the Unreal Engine only attempts to initialise its audio subsystem once during startup. Would you be willing to install PulseAudio and report back whether that makes any difference?

I just tried the PulseAudio,  the issue remains.      

then after reading the code of UE4, I have some questions about the following part:

![image](https://user-images.githubusercontent.com/33884571/109165722-d79bd780-77b6-11eb-9495-b3f738f8b27f.png)

while Initializing the audio device, will two games share the same main audio device?    

from the result, it seems like they are sharing the same audio device.

![image](https://user-images.githubusercontent.com/33884571/109167033-4cbbdc80-77b8-11eb-95e9-564d2ab4c06a.png)

i am not quite sure about this.

**Comment by adamrehn** [comment]
Yes, the two games will share the same audio output device unless you forcibly override the default behaviour using environment variables that instruct SDL to select a specific device. This was why I suggested testing PulseAudio, since its server will allow multiple applications to send output to the same audio device without causing conflicts, thus ensuring both games are able to initialise their audio subsystems correctly. The use of the same audio output device should not cause issues for the MediaIPC consumer processes though, since audio is captured inside the Unreal Engine itself prior to being sent to the audio device and is then transmitted via independent IPC objects with prefixed names.

Given that installing PulseAudio did not fix the problem you are observing, I have to assume that there's some sort of issue within MediaIPC itself that's triggered when multiple producer processes are active simultaneously. The best way to test this would be to run the two games in Docker containers, since you can vary the arguments to the [docker run](https://docs.docker.com/engine/reference/commandline/run/) command to test both two isolated IPC namespaces and one shared IPC namespace. If everything works correctly in separate IPC namespaces but the problem is reproduced when using a single IPC namespace then that would confirm that there is a bug in MediaIPC itself. The commands to do so would look something like this:

```bash
# Start a container within which to run the first game, with its own IPC namespace
docker run --rm -ti -v/path/to/CompleteRPG:/hostdir -w /hostdir 'adamrehn/ue4-runtime:latest' bash
# Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering

# Start a container within which to run the second game, with its own IPC namespace
docker run --rm -ti -v/path/to/MyProject3:/hostdir -w /hostdir 'adamrehn/ue4-runtime:latest' bash
# Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering

# ------- Once the separate IPC namespace tests are done, test a single IPC namespace: -------

# Start a container within which to run the first game, using the host system's IPC namespace
docker run --rm -ti -v/path/to/CompleteRPG:/hostdir -w /hostdir --ipc=host 'adamrehn/ue4-runtime:latest' bash
# Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering

# Start a container within which to run the second game, using the host system's IPC namespace
docker run --rm -ti -v/path/to/MyProject3:/hostdir -w /hostdir --ipc=host 'adamrehn/ue4-runtime:latest' bash
# Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering
```

**Comment by hwwxj** [comment]
> ```shell
> nullrhi
> ```

Thanks for the advice, will try this asap.
BTW,
multiple producer processes are active simultaneously which causes this issue. ----  yesterday, I tried to prove this by the following tests:
![image](https://user-images.githubusercontent.com/33884571/109294866-bf849080-7868-11eb-938f-28844591dd38.png)
(producer_example means the producer provided in the example of media-IPC)
![image](https://user-images.githubusercontent.com/33884571/109295116-2a35cc00-7869-11eb-8003-3a00c0ea0ea4.png)

the result shows all of the IPC works.         Will these tests prove that multiple producer processes can be active simultaneously?

**Comment by hwwxj** [comment]
> Yes, the two games will share the same audio output device unless you forcibly override the default behaviour using environment variables that instruct SDL to select a specific device. This was why I suggested testing PulseAudio, since its server will allow multiple applications to send output to the same audio device without causing conflicts, thus ensuring both games are able to initialise their audio subsystems correctly. The use of the same audio output device should not cause issues for the MediaIPC consumer processes though, since audio is captured inside the Unreal Engine itself prior to being sent to the audio device and is then transmitted via independent IPC objects with prefixed names.
> 
> Given that installing PulseAudio did not fix the problem you are observing, I have to assume that there's some sort of issue within MediaIPC itself that's triggered when multiple producer processes are active simultaneously. The best way to test this would be to run the two games in Docker containers, since you can vary the arguments to the [docker run](https://docs.docker.com/engine/reference/commandline/run/) command to test both two isolated IPC namespaces and one shared IPC namespace. If everything works correctly in separate IPC namespaces but the problem is reproduced when using a single IPC namespace then that would confirm that there is a bug in MediaIPC itself. The commands to do so would look something like this:
> 
> ```shell
> # Start a container within which to run the first game, with its own IPC namespace
> docker run --rm -ti -v/path/to/CompleteRPG:/hostdir -w /hostdir 'adamrehn/ue4-runtime:latest' bash
> # Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering
> 
> # Start a container within which to run the second game, with its own IPC namespace
> docker run --rm -ti -v/path/to/MyProject3:/hostdir -w /hostdir 'adamrehn/ue4-runtime:latest' bash
> # Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering
> 
> # ------- Once the separate IPC namespace tests are done, test a single IPC namespace: -------
> 
> # Start a container within which to run the first game, using the host system's IPC namespace
> docker run --rm -ti -v/path/to/CompleteRPG:/hostdir -w /hostdir --ipc=host 'adamrehn/ue4-runtime:latest' bash
> # Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering
> 
> # Start a container within which to run the second game, using the host system's IPC namespace
> docker run --rm -ti -v/path/to/MyProject3:/hostdir -w /hostdir --ipc=host 'adamrehn/ue4-runtime:latest' bash
> # Once the container starts, run the consumer process and the game, being sure to pass the `-nullrhi` flag to disable rendering
> ```

For this issue, although the audio is conflict, but we found that the video is correct in both games. So I guess maybe the MediaIPC is working normally.
and do u have any audio test tools recommended in linux that can capture the audio from the soundcard?   cause I try to use ffmpeg to capture the audio of game from soundcard, but get nothing.

BTW, may I ask that is pixel stream support the audio capture in linux?

**Comment by hwwxj** [comment]
find the reason why I cannot capture the audio of the first game. 
https://answers.unrealengine.com/questions/66918/keep-audio-from-losing-focus.html

thanks a lot!

**Comment by adamrehn** [comment]
@hwwxj I'm glad you were able to hunt down the underlying cause of your audio problems! It's good to know that this was an Unreal-related issue rather than a bug in MediaIPC.

In terms of system-level capture tools, ffmpeg would typically be my go-to as well, so there isn't really anything else I could recommend other than GUI tools like OBS which aren't of much help on a headless server.

Yes, the recently-released [4.25 version of Pixel Streaming for Linux](https://github.com/ImmortalEmperor/UnrealEngine/tree/4.25-pixelstreaming) supports audio capture under Linux.
